ops/lib.rs

```diff
@@ -0,0 +1,262 @@
+use proc_macro2::Span;
+use proc_macro2::TokenStream as TokenStream2;
+use proc_macro_crate::crate_name;
+use proc_macro_crate::FoundCrate;
+use quote::quote;
+use syn::Ident;
+
+// Identifer to the `deno_core` crate.
+//
+// If macro called in deno_core, `crate` is used.
+// If macro called outside deno_core, `deno_core` OR the renamed
+// version from Cargo.toml is used.
+fn core_import() -> TokenStream2 {
+  let found_crate =
+    crate_name("deno_core").expect("deno_core not present in `Cargo.toml`");
+
+  match found_crate {
+    FoundCrate::Itself => {
+      // TODO(@littledivy): This won't work for `deno_core` examples
+      // since `crate` does not refer to `deno_core`.
+      // examples must re-export deno_core to make this work
+      // until Span inspection APIs are stabalized.
+      //
+      // https://github.com/rust-lang/rust/issues/54725
+      quote!(crate)
+    }
+    FoundCrate::Name(name) => {
+      let ident = Ident::new(&name, Span::call_site());
+      quote!(#ident)
+    }
+  }
+}
+
+#[proc_macro_attribute]
+pub fn op(_attr: TokenStream, item: TokenStream) -> TokenStream {
+  let func = syn::parse::<syn::ItemFn>(item).expect("expected a function");
+  let name = &func.sig.ident;
+  let generics = &func.sig.generics;
+  let type_params = &func.sig.generics.params;
+  let where_clause = &func.sig.generics.where_clause;
+
+  // Preserve the original func as op_foo::call()
+  let original_func = {
+    let mut func = func.clone();
+    func.sig.ident = quote::format_ident!("call");
+    func
+  };
+
+  let core = core_import();
+
+  let v8_body = match func.sig.asyncness.is_some() {
+    true => codegen_v8_async(&core, &func),
+    false => codegen_v8_sync(&core, &func),
+  };
+
+  // Generate wrapper
+  quote! {
+    #[allow(non_camel_case_types)]
+    pub struct #name;
+
+    impl #name {
+      pub fn name() -> &'static str {
+        stringify!(#name)
+      }
+
+      pub fn v8_cb #generics () -> #core::v8::FunctionCallback #where_clause {
+        use #core::v8::MapFnTo;
+        Self::v8_func::<#type_params>.map_fn_to()
+      }
+
+      pub fn decl #generics ()  -> (&'static str, #core::v8::FunctionCallback) #where_clause {
+        (Self::name(), Self::v8_cb::<#type_params>())
+      }
+
+      #[inline]
+      #original_func
+
+      pub fn v8_func #generics (
+        scope: &mut #core::v8::HandleScope,
+        args: #core::v8::FunctionCallbackArguments,
+        mut rv: #core::v8::ReturnValue,
+      ) #where_clause {
+        #v8_body
+      }
+    }
+  }.into()
+}
+
+/// Generate the body of a v8 func for an async op
+fn codegen_v8_async(core: &TokenStream2, f: &syn::ItemFn) -> TokenStream2 {
+  let a = codegen_arg(core, &f.sig.inputs[1], "a", 2);
+  let b = codegen_arg(core, &f.sig.inputs[2], "b", 3);
+  let type_params = &f.sig.generics.params;
+
+  quote! {
+    use #core::futures::FutureExt;
+    // SAFETY: Called from Deno.core.opAsync. Which retrieves the index using opId table.
```

Comment:
Can it happen that we're unable to retrieve `opId`?

Reply:
Yes, this actually a failure mode I bumped into a will fix, if a func name isn't in the table, we should handle that JS side might impact perf but we can optimize that out with other changes